apps/data-pipeline/library-traffic-scraper/src/lib.ts

```diff
@@ -0,0 +1,201 @@
+    .filter(Boolean);
+}
+
+async function collectLocationMeta(): Promise<Record<string, LocationMeta>> {
+  const html = await fetch("https://www.lib.uci.edu/where-do-you-want-study-today").then((r) =>
+    r.text(),
+  );
+  const $ = load(html);
+
+  const scriptText = $("script")
+    .toArray()
+    .map((el) => $(el).html() ?? "")
+    .find((txt) => /let\s+locationIds\s*=\s*\[/s.test(txt));
+
+  if (!scriptText) throw new Error("Could not find <script> containing locationIds");
+
+  const locMatch = scriptText.match(/let\s+locationIds\s*=\s*\[([\s\S]*?)]/);
+  const floorMatch = scriptText.match(/let\s+floors\s*=\s*\[([\s\S]*?)]/);
+  if (!locMatch || !floorMatch) throw new Error("Unable to capture locationIds or floors array");
+
+  const ids = parseArray(locMatch[1]);
+  const codes = parseArray(floorMatch[1]);
+
+  if (ids.length !== codes.length)
+    throw new Error("locationIds and floors arrays are different lengths");
+
+  const codeToLibrary = (code: string): string => {
+    if (code.includes("GSC")) return "Gateway Study Center";
+    if (code.includes("-SL-")) return "Science Library"; // SL prefix in list
+    return "Langson Library"; // fallback
+  };
+
+  const meta: Record<string, LocationMeta> = {};
+
+  ids.forEach((id, idx) => {
+    const code = codes[idx];
+    const root = $(`#leftSide${code}`);
+
+    const primary = root.find("h2.card-title").first().text().trim();
+
+    const sub = root.find("span.subLocation").first().text().trim().replace(/\s+/g, " ");
+
+    const label = sub ? `${primary} - ${sub}` : primary || "Unknown";
+
+    meta[id] = {
+      id,
+      floorCode: code,
+      libraryName: codeToLibrary(code),
+      locationLabel: label,
+    };
+  });
+
+  return meta;
+}
+
+async function fetchLocation(id: string): Promise<RawRespOK["data"] | null> {
+  const url = `https://www.lib.uci.edu/sites/all/scripts/occuspace.php?id=${id}`;
+  try {
+    const raw = await fetch(url).then((r) => r.text());
+    const parsedOnce = JSON.parse(raw);
+    const parsed = typeof parsedOnce === "string" ? JSON.parse(parsedOnce) : parsedOnce;
+    if (parsed.data && typeof parsed.data === "object") return (parsed as RawRespOK).data;
+    console.warn(`ID ${id} responded with error: ${(parsed as RawRespErr).error}`);
+  } catch (err) {
+    console.error(`Unexpected error while fetching ID ${id}:`, err);
+    throw err;
+  }
+  return null;
+}
+
+// - "active": Library location is open — perform scrape every 15 minutes.
+// - "idle": Library is closed — perform scrape every 60 minutes.
+// - "skip": Library is nonexistent — do not scrape at all.
+type ScrapeStatus = "active" | "idle" | "skip";
+
+export function getScrapeStatus(library: "LL" | "SL" | "LGSC"): ScrapeStatus {
+  const pacificTime = new Date(
+    new Date().toLocaleString("en-US", { timeZone: "America/Los_Angeles" }),
+  );
+  const hour = pacificTime.getHours();
+  const day = pacificTime.getDay();
+
+  if (library === "LL" || library === "SL") {
```

Comment:
the comments giving day of week and time of day are great for convenience

could we add more comments to explicitly explain why those times ranges are considered "active or "idle"? could be as simple as "the library is closed at this time" or "in our observation, there tends to be much less traffic at these hours since it's nighttime/the weekend/etc"